Fix a getenv segfault due to a race

deepshah133 opened the pull request:
Summary:
* TLDR:

`getenv` is not thread safe w.r.t `setenv`. Environment variables are kept as a per-process "dictionary" by libc. `setenv` can essentially realloc the whole thing move this list to a completely different location. If there is a concurrent `getenv` happening as the same time, it is possible that it might end up reading stale memory and segfault.
`getenv` is thread safe w.r.t other `getenv`.

* Details:

Inside PTD init:
```
ProcessGroupNCCL ctor
	...
	ncclCommWatchdogThread_ =
      std::thread(&ProcessGroupNCCL::ncclCommWatchdog, this); (https://fburl.com/code/terf9ai7)
```

Inside ncclCommWatchdog thread:
```
	...
	ncclHeartbeatMonitorThread_ =
        std::thread(&ProcessGroupNCCL::heartbeatMonitor, this);  (https://fburl.com/code/fv9camg2)
    ...
```

Inside heartbeatMonitor thread:
```
	...
	std::optional<DumpPipe> dumpPipe = std::nullopt; (https://fburl.com/code/qdvahzbu)
	dumpPipe.emplace(rank_);
	...
```

Inside DumpPipe ctor (https://fburl.com/code/wvixlqcz)
```
	getCvarString
		getenv <=== SIGSEGV
```

On the main thread:

We go on to initialize NCCL:

Inside getNCCLComm, we call: `getNcclVersion` -> `initEnv` (https://fburl.com/code/j312pccu)

`initEnv` inside NCCL does this: `initEnv` -> `setEnvFile`

This guy, reads the /etc/nccl.conf file, and sets values of env variables with "setenv" (https://fburl.com/code/cq4r0y0h)
This "setenv" can race with "getenv" in heartbeatMonitor thread.

Ideally, all `setenv` should be done by a single thread before launching other threads. This diff moves getNCCLVersion before launching watchdog thread to make sure all setenvs are done beforehand.

I think we are just getting lucky that we are not hitting it in production. IIRC in fact we saw getenv segfault once in one of the large scale runs, but now I dont remember the details.

Test Plan: A lot of testing done as part of D61411062 & CI

Differential Revision: D61421292


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -781,6 +781,7 @@ ProcessGroupNCCL::ProcessGroupNCCL(
       ValueError,
       at::cuda::getNumGPUs() != 0,
       "ProcessGroupNCCL is only supported with GPUs, no GPUs found!");
+  const auto ncclVersion = getNcclVersion();
```
Nit: add a comment here explaining that this call needs to stay here bc it causes the setenv race

seemethere commented:
/easycla

malfet commented:
Would be nice to get some excerpts from backtrace embedded in PR description (SIGSEGV in getenv sounds a bit suspicious to be frank)

fduwjj reviewed (approved):
LGTM

deepshah133 committed:
Fix a getenv segfault due to a race (#133744)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/133744

* TLDR:

`getenv` is not thread safe w.r.t `setenv`. Environment variables are kept as a per-process "dictionary" by libc. `setenv` can essentially realloc the whole thing move this list to a completely different location. If there is a concurrent `getenv` happening as the same time, it is possible that it might end up reading stale memory and segfault.
`getenv` is thread safe w.r.t other `getenv`.

* Details:

Inside PTD init:
```
ProcessGroupNCCL ctor
	...
	ncclCommWatchdogThread_ =
      std::thread(&ProcessGroupNCCL::ncclCommWatchdog, this); (https://fburl.com/code/terf9ai7)
```

Inside ncclCommWatchdog thread:
```
	...
	ncclHeartbeatMonitorThread_ =
        std::thread(&ProcessGroupNCCL::heartbeatMonitor, this);  (https://fburl.com/code/fv9camg2)
    ...
```

Inside heartbeatMonitor thread:
```
	...
	std::optional<DumpPipe> dumpPipe = std::nullopt; (https://fburl.com/code/qdvahzbu)
	dumpPipe.emplace(rank_);
	...
```

Inside DumpPipe ctor (https://fburl.com/code/wvixlqcz)
```
	getCvarString
		getenv <=== SIGSEGV
```

On the main thread:

We go on to initialize NCCL:

Inside getNCCLComm, we call: `getNcclVersion` -> `initEnv` (https://fburl.com/code/j312pccu)

`initEnv` inside NCCL does this: `initEnv` -> `setEnvFile`

This guy, reads the /etc/nccl.conf file, and sets values of env variables with "setenv" (https://fburl.com/code/cq4r0y0h)
This "setenv" can race with "getenv" in heartbeatMonitor thread.

Ideally, all `setenv` should be done by a single thread before launching other threads. This diff moves getNCCLVersion before launching watchdog thread to make sure all setenvs are done beforehand.

I think we are just getting lucky that we are not hitting it in production. IIRC in fact we saw getenv segfault once in one of the large scale runs, but now I dont remember the details.

Example backtrace:
```
__GI___pthread_kill at /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/nptl/pthread_kill.c:46
__GI___pthread_kill at /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/nptl/pthread_kill.c:62
__GI_raise at /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/sysdeps/posix/raise.c:26
__restore_rt at /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/sysdeps/unix/sysv/linux/libc_sigaction.c:13
__GI_getenv at /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/stdlib/getenv.c:84
c10d::getCvarString(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>>> const&, char const*) at Utils.hpp:106
c10d::DumpPipe::DumpPipe(int) at ProcessGroupNCCL.hpp:168
c10d::ProcessGroupNCCL::heartbeatMonitor() at /opt/rh/gcc-toolset-11/root/usr/include/c++/11/bits/stl_construct.h:119
c10d::ProcessGroupNCCL::heartbeatMonitor() at /opt/rh/gcc-toolset-11/root/usr/include/c++/11/optional:251
c10d::ProcessGroupNCCL::heartbeatMonitor() at /opt/rh/gcc-toolset-11/root/usr/include/c++/11/optional:419
c10d::ProcessGroupNCCL::heartbeatMonitor() at /opt/rh/gcc-toolset-11/root/usr/include/c++/11/optional:875
c10d::ProcessGroupNCCL::heartbeatMonitor() at ProcessGroupNCCL.cpp:1259
execute_native_thread_routine at unknown:-1
start_thread at /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/nptl/pthread_create.c:434
__clone3 at /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```

Test Plan: A lot of testing done as part of D61411062 & CI

Reviewed By: cenzhaometa

Differential Revision: D61421292
